closeio_api/__init__.py

```diff
@@ -44,110 +55,202 @@ def __init__(self, base_url, api_key=None, tz_offset=None,
+        func = self._prepare_request if self.async else self._dispatch
+        return func('post', endpoint+'/', **kwargs)
 
     def put(self, endpoint, data, **kwargs):
+        """Send (sync client) or prepare (async client) a PUT request to
+        a given endpoint, for example:
+
+        >>> api.put('lead/SOME_LEAD_ID', {'name': 'New Name'})
+        {
+            'name': 'New Name'
+            # ... rest of the response omitted for brevity
+        }
+        """
         kwargs.update({'data': data})
-        return self.dispatch('put', endpoint+'/', **kwargs)
+        func = self._prepare_request if self.async else self._dispatch
+        return func('put', endpoint+'/', **kwargs)
 
     def delete(self, endpoint, **kwargs):
-        return self.dispatch('delete', endpoint+'/', **kwargs)
+        """Send (sync client) or prepare (async client) a DELETE request
+        to a given endpoint, for example:
+
+        >>> api.delete('lead/SOME_LEAD_ID')
+        {}
+        """
+        func = self._prepare_request if self.async else self._dispatch
+        return func('delete', endpoint+'/', **kwargs)
 
     # Only for async requests
     def map(self, reqs, max_retries=None):
+        """Execute a batch of asynchronous requests concurrently. For
+        example:
+
+        >>> reqs = []
+        >>> reqs.append(api.post('lead', {'name': 'New Lead'}))
+        >>> reqs.append(api.post('lead', {'name': 'Another New Lead'}))
+        >>> api.map(reqs)
+        [
+            # list of dicts containing successful API responses and APIError
+            # objects for failed requests.
+        ]
+        """
+        if not self.async:
+            raise NotImplementedError('map can only be used in an async Client')
+
+        if len(reqs) > MAX_CONCURRENT_REQUESTS:
+            raise ValueError(
+                'Too many concurrent requests ({}). You can send up to {} in '
+                'a single batch.'.format(len(reqs), MAX_CONCURRENT_REQUESTS)
+            )
+
         if max_retries is None:
             max_retries = self.max_retries
+
         # TODO
         # There is no good way of catching or dealing with exceptions that are
         # raised during the request sending process when using map or imap.
         # When this issue is closed:
         # https://github.com/kennethreitz/grequests/pull/15
         # modify this method to repeat only the requests that failed because of
         # connection errors
-        if self.async:
-            import grequests
-            responses = [(
+
+        responses = [(
+            response.json() if response.ok else APIError(response)
+        ) for response in self.requests.map(reqs)]
+
+        # retry the api calls that failed until they succeed or the
+        # max_retries limit is reached
+        retries = 0
+        while True and retries < max_retries:
+            n_errors = sum([int(isinstance(response, APIError))
```

Comment:
`ValidationError` inherits `APIError`. We should make sure it doesn't get retried though.